uberfire-workbench/uberfire-workbench-processors/src/main/java/org/uberfire/annotations/processors/ClientEditorActivityGenerator.java

```diff
@@ -0,0 +1,304 @@
+import freemarker.template.TemplateException;
+import org.uberfire.annotations.processors.exceptions.GenerationException;
+import org.uberfire.annotations.processors.facades.ClientAPIModule;
+
+/**
+ * A source code generator for Activities
+ */
+public class ClientEditorActivityGenerator extends AbstractGenerator {
+
+    @Override
+    public StringBuffer generate(final String packageName,
+                                 final PackageElement packageElement,
+                                 final String className,
+                                 final Element element,
+                                 final ProcessingEnvironment processingEnvironment) throws GenerationException {
+
+        final Messager messager = processingEnvironment.getMessager();
+        messager.printMessage(Kind.NOTE,
+                              "Starting code generation for [" + className + "]");
+
+        //Extract required information
+        final TypeElement classElement = (TypeElement) element;
+
+        final String annotationName = ClientAPIModule.getWorkbenchClientEditorClass();
+        final boolean isDynamic = ClientAPIModule.getWbEditorIsDynamicValueOnClass(classElement);
+
+        String identifier = null;
+
+        for (final AnnotationMirror am : classElement.getAnnotationMirrors()) {
+            if (annotationName.equals(am.getAnnotationType().toString())) {
+                for (Map.Entry<? extends ExecutableElement, ? extends AnnotationValue> entry : am.getElementValues().entrySet()) {
+                    AnnotationValue aval = entry.getValue();
+                    if ("identifier".equals(entry.getKey().getSimpleName().toString())) {
+                        identifier = aval.getValue().toString();
+                    }
+                }
+                break;
+            }
+        }
+
+        final String beanActivatorClass = GeneratorUtils.getBeanActivatorClassName(classElement,
+                                                                                   processingEnvironment);
+
+        final ExecutableElement onStartupMethod = GeneratorUtils.getOnStartupMethodForNonEditors(classElement,
+                                                                                                 processingEnvironment);
+
+        final String onStartup0ParameterMethodName;
+        final String onStartup1ParameterMethodName;
+        if (onStartupMethod == null) {
+            onStartup0ParameterMethodName = null;
+            onStartup1ParameterMethodName = null;
+        } else if (onStartupMethod.getParameters().isEmpty()) {
+            onStartup0ParameterMethodName = onStartupMethod.getSimpleName().toString();
+            onStartup1ParameterMethodName = null;
+        } else {
+            onStartup0ParameterMethodName = null;
+            onStartup1ParameterMethodName = onStartupMethod.getSimpleName().toString();
+        }
+
+        final String onMayCloseMethodName = GeneratorUtils.getOnMayCloseMethodName(classElement,
```

Comment:
I think you could construct map `Map<String, ExecutableElement>` and then the following parts of code could just process this map in cycle. The map could be created in separate method for better readability.

Reply:
Sorry, I didn't understand exactly how this could be changed.  I think the code for this was based on `EditorActivityGenerator`. I could make some changes if you want. One example is this if/else to determine if there's a startup method and if it has params, it looks like a boolean logic to me, where 1 is the method name and 0 null, so:

```
startup exists    startup has params    noParam   withParam
0                                          0                 0         0
1                                          0                 1         0  
0                                          1                 0         0                
1                                          1                 0         1
```

So the resulting code:

~~~
        final String onStartup0ParameterMethodName = (onStartupMethod != null && onStartupMethod.getParameters().isEmpty()) ? 
                                                               onStartupMethod.getSimpleName().toString() 
                                                               : null;
        final String onStartup1ParameterMethodName = (onStartupMethod != null && !onStartupMethod.getParameters().isEmpty()) ? 
                                                                    onStartupMethod.getSimpleName().toString() 
                                                                    : null;
~~~

which looks worse.

The whole method could be condensed in a map, and late a for in the map fill the methods accordingly, but I would vote to avoid change it because it is, as I said, based on EditorActivityGenerator. What do you think?

Reply:
I meant something like this
```
Map<String, String> methodNameMap = ... // or ExecutableElement and you could call setContentMethod.getSimpleName().toString(); on each value of the map later
map.put("onCloseMethodName", onCloseMethodName);
```
Following code would  be better readable.